### PR TITLE
ASVideoPlayerNode Beta

### PR DESCRIPTION
**Pull request by Eke**
##### What is this?!
`ASVideoPlayerNode` is `ASVideoNode` UI wrapper. It is highly customisable and gives developer ability to build UI above `ASVideoNode`. It has default template but gives developer ability to build fully custom UI with delegate methods

#### So what about `ASVideoNode`?
`ASVideoNode` will remain as video playback core. So `ASVideoPlayerNode` will be responsible for UI, `ASVideoNode` for playback

##### videoPlayerNodeNeededControls:
Asks delegate which controls does it need in certain `ASVideoPlayerNode`.  By default all controls are enabled but delegate can override this and use only `ASVideoPlayerNodeControlTypePlaybackButton`.
```
- (NSArray *)videoPlayerNodeNeededControls:(ASVideoPlayerNode *)videoPlayer
{
  return @[ @(ASVideoPlayerNodeControlTypePlaybackButton) ];
}
```
##### videoPlayerNodeLayoutSpec:forControls:forMaximumSize:
Asks delegate for `ASLayoutSpec` which will be used to layout  certain `ASVideoPlayerNode`. With this method developer can  whole layout for `ASVideoPlayerNode` except `ASVideoNode`, it's locked.

```
- (ASLayoutSpec *)videoPlayerNodeLayoutSpec:(ASVideoPlayerNode *)videoPlayer forControls:(NSDictionary *)controls forMaximumSize:(CGSize)maxSize
{
  NSMutableArray *bottomControls = [[NSMutableArray alloc] init];

  ASDisplayNode *playbackButtonNode = controls[@(ASVideoPlayerNodeControlTypePlaybackButton)];

  if (playbackButtonNode) {
    [bottomControls addObject:playbackButtonNode];
  }

  ASLayoutSpec *spacer = [[ASLayoutSpec alloc] init];
  spacer.flexGrow = YES;

  ASStackLayoutSpec *controlbarSpec = [ASStackLayoutSpec stackLayoutSpecWithDirection:ASStackLayoutDirectionHorizontal
  spacing:10.0
  justifyContent:ASStackLayoutJustifyContentStart
  alignItems:ASStackLayoutAlignItemsCenter
  children:bottomControls];
  controlbarSpec.alignSelf = ASStackLayoutAlignSelfStretch;

  UIEdgeInsets insets = UIEdgeInsetsMake(10.0, 10.0, 10.0, 10.0);

  ASInsetLayoutSpec *controlbarInsetSpec = [ASInsetLayoutSpec insetLayoutSpecWithInsets:insets child:controlbarSpec];

  controlbarInsetSpec.alignSelf = ASStackLayoutAlignSelfStretch;

  ASStackLayoutSpec *mainVerticalStack = [ASStackLayoutSpec stackLayoutSpecWithDirection:ASStackLayoutDirectionVertical
  spacing:0.0
  justifyContent:ASStackLayoutJustifyContentStart
  alignItems:ASStackLayoutAlignItemsStart
  children:@[ spacer, controlbarInsetSpec ]];


  return mainVerticalStack;
}
```

In this example we position play button at the bottom-left corner of `ASVideoPlayerNode`. `controls` dictionary contains only same elements which were returned by `videoPlayerNodeNeededControls:`

##### videoPlayerNodeTimeLabelAttributes:timeLabelType
Asks delegate for `NSAttributedString` options for duration and elapsed time labels. Developer can control everything, starting from color ending with `UIFont`

##### videoPlayerNode:timeStringForTimeLabelType:forTime:
Gives delegate duration or current playback `CMTime` and asks to format it in desired way.

###### videoPlayerNodePlaybackButtonTint:

Also there are all `ASVideoNode` delegate methods available :
###### videoPlayerNodeDidPlayToEnd:

Example project is not finished yet, will try to finish it asap

### CODE REVIEW

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+                                                                                 spacing:0.0
+                                                                          justifyContent:ASStackLayoutJustifyContentStart
+                                                                              alignItems:ASStackLayoutAlignItemsStart
+                                                                                children:@[spacer,controlbarInsetSpec]];
+
+  return mainVerticalStack;
+}
+
+#pragma mark - Properties
+- (id<ASVideoPlayerNodeDelegate>)delegate{
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Bracket on new line

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+}
+
+- (void)videoPlaybackDidFinish:(ASVideoNode *)videoNode
+{
+  if (_delegateFlags.delegateVideoNodePlaybackDidFinish) {
+    [_delegate videoPlayerNodeDidPlayToEnd:self];
+  }
+}
+
+- (void)videoNodeWasTapped:(ASVideoNode *)videoNode
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: didTapVideoNode

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+    if (videoNode.playerState == ASVideoNodePlayerStatePlaying) {
+      [videoNode pause];
+    } else {
+      [videoNode play];
+    }
+  }
+}
+
+#pragma mark - Actions
+- (void)playbackButtonTapped:(ASControlNode*)node
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: didTapPlaybackButton

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+{
+  _isSeeking = YES;
+}
+
+- (void)endedSeek
+{
+  _isSeeking = NO;
+}
+
+- (void)changedSeekValue:(UISlider*)slider
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: seekTimeDidChange

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+
+      return slider;
+    }];
+
+    _scrubberNode.flexShrink = YES;
+
+    [_cachedControls setObject:_scrubberNode forKey:@(ASVideoPlayerNodeControlTypeScrubber)];
+  }
+
+  [self addSubnode:_scrubberNode];
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: If you enable implicit hierarchy management, you won't need to add the subnode manually.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+
+  _asset = asset;
+  _disableControls = NO;
+
+  [self privateInit];
+
+  return self;
+}
+
+- (void)privateInit
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: rename to _init

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+    [_videoNode play];
+  }
+}
+
+- (void)beganSeek
+{
+  _isSeeking = YES;
+}
+
+- (void)endedSeek
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: These two methods should be present tense. Ie. beginSeek and endSeek.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+  _videoNode.muted = _muted;
+}
+
+- (void)setPeriodicTimeObserverTimescale:(int32_t)periodicTimeObserverTimescale
+{
+  _periodicTimeObserverTimescale = periodicTimeObserverTimescale;
+  _videoNode.periodicTimeObserverTimescale = _periodicTimeObserverTimescale;
+}
+
+- (NSString*)gravity
```
Comment (levi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Space between *

**Comment by Eke** [comment]
@levi thanks for comments, will fix in few hours

**Comment by levi** [comment]
@Eke something to think about is extracting the `showSpinner` method in `ASVideoNode` into this new player. The original video node shouldn't be responsible for displaying any sort of UI, rather delegating it to higher abstracted nodes that can easily be customized (like ASVideoPlayerNode).

**Comment by Eke** [comment]
Hello @levi, yes, all UI elements will be removed from `ASVideoNode`, it's on my todo list

**Comment by Xinchi** [comment]
Could you fix this: 


❌  /Users/travis/build/facebook/AsyncDisplayKit/examples/Videos/Sample/ViewController.m:41:18: property 'playButton' not found on object of type 'ASVideoNode *'
  simonVideoNode.playButton = self.playButton;
                 ^

**Comment by Eke** [comment]
@Xinchi we must remove all `ASVideoNode` examples to fix this error.

File: AsyncDisplayKit/ASVideoPlayerNode.h [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+#if TARGET_OS_IOS
+#import <AsyncDisplayKit/AsyncDisplayKit.h>
+//#import <AsyncDisplayKit/ASThread.h>
+//#import <AsyncDisplayKit/ASVideoNode.h>
+//#import <AsyncDisplayKit/ASDisplayNode+Subclasses.h>
+
+@class AVAsset;
+@protocol ASVideoPlayerNodeDelegate;
+
+typedef enum {
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.h]: I'm a little concerned about the flexibility of adding additional nodes with this approach.  An example is that what if the client needs to add the sound control?  Or additional app-specific control components?

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 10]
```diff
@@ -0,0 +1,729 @@
...
+    _delegateFlags.delegateVideoNodeShouldChangeState = [_delegate respondsToSelector:@selector(videoPlayerNode:shouldChangeVideoNodeStateTo:)];
+    _delegateFlags.delegateTimeLabelAttributedString = [_delegate respondsToSelector:@selector(videoPlayerNode:timeStringForTimeLabelType:forTime:)];
+    _delegateFlags.delegatePlaybackButtonTint = [_delegate respondsToSelector:@selector(videoPlayerNodePlaybackButtonTint:)];
+    _delegateFlags.delegateDidTapVideoPlayerNode = [_delegate respondsToSelector:@selector(didTapVideoPlayerNode:)];
+  }
+}
+
+- (void)setDisableControls:(BOOL)disableControls
+{
+  _disableControls = disableControls;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: do a check: if (_disableControls != disableControls)

File: AsyncDisplayKit/ASVideoPlayerNode.h [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface ASVideoPlayerNode : ASDisplayNode
+
+@property (nullable, atomic, weak) id<ASVideoPlayerNodeDelegate> delegate;
+
+@property (nonatomic,assign,readonly) CMTime duration;
+
+@property (nonatomic, assign) BOOL disableControls;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.h]: Rename to controlsDisabled

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 12]
```diff
@@ -0,0 +1,729 @@
...
+{
+  if (_delegateFlags.delegateDidTapVideoPlayerNode) {
+    [_delegate didTapVideoPlayerNode:self];
+  } else {
+    [self manageVideoNodePlayback];
+  }
+}
+
+#pragma mark - Actions
+- (void)manageVideoNodePlayback
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: rename this method to togglePlayPause

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 13]
```diff
@@ -0,0 +1,729 @@
...
+
+  if (_neededControls == nil) {
+    _neededControls = [self createControlElementArray];
+  }
+
+  if (_cachedControls == nil) {
+    _cachedControls = [[NSMutableDictionary alloc] init];
+  }
+
+  for (int i = 0; i < _neededControls.count; i++) {
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: for (id object in array) {
}

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 14]
```diff
@@ -0,0 +1,729 @@
...
+- (void)didLoad
+{
+  [super didLoad];
+  {
+    ASDN::MutexLocker l(_videoPlayerLock);
+    [self createControls];
+  }
+}
+
+- (NSArray*)createControlElementArray
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: nit: space between NSArray and *

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 15]
```diff
@@ -0,0 +1,729 @@
...
+
+  return @[ @(ASVideoPlayerNodeControlTypePlaybackButton),
+            @(ASVideoPlayerNodeControlTypeElapsedText),
+            @(ASVideoPlayerNodeControlTypeScrubber),
+            @(ASVideoPlayerNodeControlTypeDurationText) ];
+}
+
+- (void)addObservers
+{
+
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Empty method?

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 16]
```diff
@@ -0,0 +1,729 @@
...
+}
+
+- (instancetype)initWithAsset:(AVAsset *)asset
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+
+  _asset = asset;
+  _disableControls = NO;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Don't need this line.  Default is NO.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 17]
```diff
@@ -0,0 +1,729 @@
...
+}
+
+- (void)_init
+{
+
+  _defaultControlsColor = [UIColor whiteColor];
+  _cachedControls = [[NSMutableDictionary alloc] init];
+
+  _videoNode = [[ASVideoNode alloc] init];
+  _videoNode.asset = _asset;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Setting the asset here will cause the ASVideoNode to start loading the asset data asynchronously.   There are cases where the client won't perform this expensive operation when the node is initialized.  We might want to move this to when the node becomes visible.  

Or even better, we can let the client decide whether to "preload" the asset.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 18]
```diff
@@ -0,0 +1,729 @@
...
+}
+
+- (void)cleanCachedControls
+{
+  [_cachedControls removeAllObjects];
+
+  _playbackButtonNode = nil;
+  _elapsedTextNode = nil;
+  _durationTextNode = nil;
+  _scrubberNode = nil;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: What about the controls that the client specified earlier?

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 19]
```diff
@@ -0,0 +1,729 @@
...
+- (void)updateDurationTimeLabel
+{
+  NSString *formatedDuration = [self timeStringForCMTime:_duration forTimeLabelType:ASVideoPlayerNodeControlTypeDurationText];
+  _durationTextNode.attributedString = [self timeLabelAttributedStringForString:formatedDuration forControlType:ASVideoPlayerNodeControlTypeDurationText];
+}
+
+- (void)updateElapsedTimeLabel:(NSTimeInterval)seconds
+{
+  NSString *formatedDuration = [self timeStringForCMTime:CMTimeMakeWithSeconds( seconds, _videoNode.periodicTimeObserverTimescale ) forTimeLabelType:ASVideoPlayerNodeControlTypeElapsedText];
+  _elapsedTextNode.attributedString = [self timeLabelAttributedStringForString:formatedDuration forControlType:ASVideoPlayerNodeControlTypeElapsedText];
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Type here: formatedDuration -> formattedDuration

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 20]
```diff
@@ -0,0 +1,729 @@
...
+
+#pragma mark - ASVideoNodeDelegate
+- (void)videoNode:(ASVideoNode *)videoNode willChangePlayerState:(ASVideoNodePlayerState)state toState:(ASVideoNodePlayerState)toState
+{
+  if (_delegateFlags.delegateVideoNodeWillChangeState) {
+    [_delegate videoPlayerNode:self willChangeVideoNodeState:state toVideoNodeState:toState];
+  }
+
+  if (toState == ASVideoNodePlayerStateReadyToPlay && _durationTextNode) {
+    _duration = _videoNode.currentItem.duration;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: My hunch is telling me we want this duration to be captured even without the existence of the durationTextNode.  Aka the client want to log the duration of the video, but the actual durationTextNode isn't needed on the UI.

File: AsyncDisplayKit/ASVideoNode.h [hunk 21]
```diff
@@ -15,8 +15,10 @@
 typedef enum {
   ASVideoNodePlayerStateUnknown,
   ASVideoNodePlayerStateInitialLoading,
-  ASVideoNodePlayerStateLoading,
+  ASVideoNodePlayerStateReadyToPlay,
+  ASVideoNodePlayerStatePlaybackLikelyToKeepUpButNotPlaying,
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoNode.h]: This is a nice addition for stall cases!

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 22]
```diff
@@ -0,0 +1,729 @@
...
+
+  if (toState == ASVideoNodePlayerStatePlaying) {
+    _playbackButtonNode.buttonType = ASDefaultPlaybackButtonTypePause;
+    [self removeSpinner];
+  } else if (toState != ASVideoNodePlayerStatePlaybackLikelyToKeepUpButNotPlaying && toState != ASVideoNodePlayerStateReadyToPlay) {
+    _playbackButtonNode.buttonType = ASDefaultPlaybackButtonTypePlay;
+  }
+
+  if (toState == ASVideoNodePlayerStateLoading || toState == ASVideoNodePlayerStateInitialLoading) {
+    [self showSpinner];
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Nice moving the UI over!

File: AsyncDisplayKit/ASVideoPlayerNode.h [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+#pragma mark - Spinner delegate methods
+- (UIColor *)videoPlayerNodeSpinnerTint:(ASVideoPlayerNode *)videoPlayer;
+
+#pragma mark - Playback button delegate methods
+- (UIColor *)videoPlayerNodePlaybackButtonTint:(ASVideoPlayerNode *)videoPlayer;
+
+
+#pragma mark ASVideoNodeDelegate proxy methods
+/**
+ * @abstract Delegate method invoked when ASVideoPlayerNode playback time is taped.
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.h]: playback time is tapped?  (watch for typos)

File: AsyncDisplayKit/ASVideoPlayerNode.h [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+ * @abstract Delegate method invoked when ASVideoNode changes state.
+ * @param videoPlayerNode The ASVideoPlayerNode whose ASVideoNode is changing state.
+ * @param state ASVideoNode state before this change.
+ * @param toSate ASVideoNode new state.
+ * @discussion This method is called after each state change
+ */
+- (void)videoPlayerNode:(ASVideoPlayerNode *)videoPlayer willChangeVideoNodeState:(ASVideoNodePlayerState)state toVideoNodeState:(ASVideoNodePlayerState)toState;
+
+/**
+ * @abstract Delegate method is invoked when ASVideoNode decides to change state.
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.h]: Need more explanation of what this method does (aka not allowing certain state change..etc)

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 25]
```diff
@@ -0,0 +1,729 @@
...
+{
+  if (_delegateFlags.delegateVideoNodeShouldChangeState) {
+    return [_delegate videoPlayerNode:self shouldChangeVideoNodeStateTo:state];
+  }
+  return YES;
+}
+
+- (void)videoNode:(ASVideoNode *)videoNode didPlayToTimeInterval:(NSTimeInterval)timeInterval
+{
+  //TODO: ask Max about CMTime problem in ASVideoNode Header file
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: We can stick with NSTimeInterval, it's fine.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 26]
```diff
@@ -0,0 +1,729 @@
...
+  if (_scrubberNode) {
+    [(UISlider*)_scrubberNode.view setValue:( timeInterval / CMTimeGetSeconds(_duration) ) animated:NO];
+  }
+}
+
+- (void)videoDidPlayToEnd:(ASVideoNode *)videoNode
+{
+  if (_delegateFlags.delegateVideoNodePlaybackDidFinish) {
+    [_delegate videoPlayerNodeDidPlayToEnd:self];
+  }
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Need some default action here: Seek to the first frame, or repeat automatically if the flag is true, or show the play button on the video node.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 27]
```diff
@@ -0,0 +1,729 @@
...
+}
+
+#pragma mark - Actions
+- (void)manageVideoNodePlayback
+{
+  if (_videoNode.playerState == ASVideoNodePlayerStatePlaying) {
+    [_videoNode pause];
+  } else {
+    [_videoNode play];
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Need to make sure this method is robust in stall cases.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 28]
```diff
@@ -0,0 +1,729 @@
...
+
+#pragma mark - Properties
+- (id<ASVideoPlayerNodeDelegate>)delegate
+{
+  return _delegate;
+}
+
+- (void)setDelegate:(id<ASVideoPlayerNodeDelegate>)delegate
+{
+  _delegate = delegate;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Add     
if (delegate == _delegate) {
        return;
    }

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 29]
```diff
@@ -0,0 +1,729 @@
...
+- (void)setShouldAutoplay:(BOOL)shouldAutoplay
+{
+  _shouldAutoplay = shouldAutoplay;
+  _videoNode.shouldAutoplay = _shouldAutoplay;
+}
+
+- (void)setShouldAutorepeat:(BOOL)shouldAutorepeat
+{
+  _shouldAutorepeat = shouldAutorepeat;
+  _videoNode.shouldAutorepeat = YES;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: = YES? That is dictatorship man.

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 30]
```diff
@@ -0,0 +1,729 @@
...
+  if (_disableControls && _cachedControls.count > 0) {
+    [self removeControls];
+  } else if (!_disableControls) {
+    [self createControls];
+  }
+}
+
+- (void)setShouldAutoplay:(BOOL)shouldAutoplay
+{
+  _shouldAutoplay = shouldAutoplay;
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: Check 
if (_shoudlAutoplay == shouldAutoplay)

File: AsyncDisplayKit/ASVideoPlayerNode.mm [hunk 31]
```diff
@@ -0,0 +1,729 @@
...
+  _videoNode.shouldAutoplay = _shouldAutoplay;
+}
+
+- (void)setShouldAutorepeat:(BOOL)shouldAutorepeat
```
Comment (Xinchi) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: nit: Upper case Repeat.

**Comment by Xinchi** [comment]
@Eke Great change! The PR looks good to me if you could address the PR comments.  One concern I have about this design is the flexibility in adding custom control elements such as the sound button and etc.  Do you have a plan for that?

File: AsyncDisplayKit/ASVideoPlayerNode.mm (hunk 26 above)
Comment (Eke) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: No need to do default actions in `ASVideoPlayerNode` `videoDidPlayToEnd ` method , `ASVideoNode` handles it, seeks to zero second or start playback again if repeat flag is true.

**Comment by Eke** [comment]
Thanks for review @Xinchi ! fixed almost all issues, just want to discuss some comments with you in Slack before pushing changes.

File: AsyncDisplayKit/ASVideoPlayerNode.mm (hunk 17 above)
Comment (Eke) [inline AsyncDisplayKit/ASVideoPlayerNode.mm]: fixed, but need to discuss

**Comment by aaronschubert0** [comment]
Looking forward to this hitting master. Most dev's will make their own UI on top of ASVideoNode so it's nice that this will separate it out.

**Comment by appleguy** [comment]
Someone do give me a heads up when this is fully reviewed and ready to land -- in Cambodia right now, so I won't get around to a review detailed enough to make the decision, but trust many in the community to do so.

**Comment by Xinchi** [comment]
Is there any conflicts when merging the master branch?

**Comment by Eke** [comment]
@Xinchi , yes there was one conflict caused by this PR #1680. resolved

**Comment by Xinchi** [comment]
Hi @appleguy , this PR looks good to me now after a couple of iterations. If there's no objections from others, would you mind merging it?

**Comment by Eke** [comment]
@appleguy there is just one thing, example app is not finished yet. If this is problem, i can finish it on weekend

**Comment by appleguy** [comment]
Lots of work here, thanks everyone for reviewing so actively and for @eke leading the change!

**Comment by wy19901227** [comment]
@Eke  ASVidepPlayerNode is very not smooth

**Comment by Eke** [comment]
@wy19901227 Hello, can you give more detailed information? Maybe i can help you

**Comment by wy19901227** [comment]
@Eke the tableview when scroll it will be very slack（not smooth）

**Comment by Eke** [comment]
@wy19901227 demo is not finished yet, will take a look again in few days. Thanks for feedback !

**Pull request merged**